Add syntax vignette (#137)
r-lib/tidyselect: vignettes/syntax.Rmd

---

> title: "The tidyselect syntax"

Reviewer: I think you need a name more like "tidyselect details" to make it clear that this is aimed at the package author, not the the data scientist.

Author: This is now `Technical description of tidyselect`

output: rmarkdown::html_vignette
vignette: >

>   %\VignetteIndexEntry{syntax}

Reviewer: Don't you need a title here?

  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
options(
  tibble.print_min = 4,
  tibble.print_max = 4
)
options(
  crayon.enabled = FALSE
)
```

This is a technical description of the tidyselect syntax.

```{r setup}
library(tidyselect)
library(magrittr)

# FIXME: `select_pos()` should be exported as `eval_select()`
select_pos <- function(data, ...) {
  tidyselect:::select_pos(data, !!rlang::expr(c(...)))
}

# For better printing
mtcars <- tibble::as_tibble(mtcars)
iris <- tibble::as_tibble(iris)
```

## Sets of variables

The tidyselect syntax is all about __sets__ of variables, internally

> represented by integer vectors of __positions__. `c(1L, 2L, 1L)`

Reviewer: For example, `c(1L, 2L)` represents the first and second variables, as does `c(1L, 2L, 1L)`.

represents the set of the first and second variables.

If a vector of positions contains duplicates, they are normally

> treated as the same element since they represent sets. An exception to

Reviewer: ```suggestion
treated as the same element, since they represent sets. An exception to
```

this occurs with named elements whose names differ. If the names don't
match, they are treated as different elements in order to allow
renaming a variable to multiple names (see section on Renaming
variables).

The syntax of tidyselect is generally designed for set combination.
For instance, `c(foo(), bar())` represents the union of the variables
in `foo()` and those in `bar()`.

### Bare names

Within _data-expressions_ (see Evaluation section), bare names
represent their own positions, i.e. a set of size 1. The following
expressions are equivalent:

```{r}
mtcars %>% select_pos(mpg:hp, -cyl, vs)

mtcars %>% select_pos(1:4, -2, 8)
```

> 

Reviewer: Maybe it's worth a mention of the other arithmetic operations somewhere? i.e. `mtcars %>% select_pos(mpg * wt)` fails because ...

```{r}
iris %>% select_pos(starts_with("Sepal") & !ends_with("Width"))
```

### Dots, `c()`, and unary `-`

tidyselect functions can take dots like `dplyr::select()`, or a named
argument like `tidyr::pivot_longer()`. In the latter case, the dots
syntax is accessible via `c()`. In fact `...` syntax is implemented
through `c(...)` and is thus completely equivalent.

```{r}
mtcars %>% select_pos(mpg, disp:hp)

mtcars %>% select_pos(c(mpg, disp:hp))
```

Dots and `c()` are syntax for:

* Set union or set difference
* Renaming variables

Non-negative inputs are recursively joined with `union()`. The
precedence is left-associative, just like with boolean operators.
These expressions are all syntax for _set union_:

```{r}
iris %>% select_pos(starts_with("Sepal"), ends_with("Width"), Species)

iris %>% select_pos(starts_with("Sepal") | ends_with("Width") | Species)

iris %>% select_pos(union(union(starts_with("Sepal"), ends_with("Width")), 5L))
```

Unary `-` is normally syntax for _set difference_:

```{r}
iris %>% select_pos(starts_with("Sepal"), -ends_with("Width"), -Sepal.Length)

iris %>% select_pos(setdiff(setdiff(starts_with("Sepal"), ends_with("Width")), 1L))
```

> If the first `...` or `c()` input is negative, an implicit

Reviewer: What about `c(mpg, c(-vs))`?

Author: I added:

````markdown
Each level of `c()` is independent. In particular, a nested `c()`
starting with `-` always stands for set complement:

```{r}
iris %>% select_pos(c(starts_with("Sepal"), -Sepal.Length))

iris %>% select_pos(c(starts_with("Sepal"), c(-Sepal.Length)))
```

In boolean terms, these expressions are equivalent to:

```{r}
iris %>% select_pos(starts_with("Sepal") & !Sepal.Length)

iris %>% select_pos(starts_with("Sepal") | !Sepal.Length)
```
````

are pasted together with a `...` separator. Otherwise the outer
names is propagated to the selected elements according to the
following rules:

- With data frames, a numeric suffix is appended because columns
  must be uniquely named.

- With normal vectors, the name is simply assigned to all selected
  inputs.

Unnamed elements match any names:

- `a | c(foo = a)` is equivalent to `c(foo = a)`.
- `a & c(foo = a)` is equivalent to `c(foo = a)`.
- `c(foo = a) & c(bar = a)` is equivalent to `c()`.
- `c(foo = a) | c(bar = a)` is equivalent to `c(foo = a, bar = a)`.

Because unnamed elements match any named ones, it is possible to
select multiple elements and rename one of them:

```{r}
iris %>% select_pos(!Species, foo = Sepal.Width)
```

### Predicate functions

Predicate function objects can be supplied as input. They are applied
to all elements of the input, and should return `TRUE` or `FALSE` to
indicate inclusion. Predicates in data-expressions are effectively
expanded to the set of variables that they represent:

```{r}
iris %>% select_pos(is.numeric)

iris %>% select_pos(is.factor)

iris %>% select_pos(is.numeric | is.factor)

iris %>% select_pos(is.numeric & is.factor)
```

> 

Reviewer: Mention ambiguity with variable names here?

Reviewer: Oh, just mention that you'll come back to it later

```{r}
x <- data.frame(x = 1:3, y = 4:6, z = 7:9)

# `ncol(x)` is an env-expression, so `x` represents the data frame in
# the environment rather than the column in the data frame
x %>% select_pos(2:ncol(x))
```

If you have variable names in a character vector, it is safe to refer
to the env-var containing the names with `all_of()` because it is an
env-expression:

```{r}
y <- c("y", "z")
x %>% select_pos(all_of(y))
```

Note that currently, env-vars are still allowed in some
data-expressions, for compatibility. However this is in the process of
being deprecated and you should see a note recommending to use
`all_of()` instead. This note will become a deprecation warning in the
future, and then an error.

```{r}
mtcars %>% select_pos(cyl_pos)
```

### Predicate functions within data-expressions

There is one exception to the unambiguous lookup rules in data
expressions. If a data-var is not found in the data frame, tidyselect
tries to find a predicate function in the environment.

This introduces an ambiguity when a predicate function has the same
name as a column. This is a conscious design decision:

- It makes it easier to use predicate functions interactively.

> - Predicate functions and data frame columns should normally have

Reviewer: ```suggestion
- The names of predicate functions and data frame columns rarely overlap
```
